url5997ccb760_JANNLab_JANNLab_tgz-pJ8-de_jannlab_examples_recurrent_SequentialXorExampleJ8/src/de/jannlab/data/SampleTools.java

```diff
@@ -0,0 +1,352 @@
+    }
+    
+    public static void readCSV(final String filename, final SampleSet set) throws IOException {
+        //
+        File file             = new File(filename);
+        BufferedReader reader = new BufferedReader(new FileReader(file));
+        //
+        while (reader.ready()) {
+            //
+            String line1 = null;
+            String line2 = null;
+            boolean grab1 = true;
+            boolean grab2 = true;
+            //
+            // grab first line:
+            //
+            while (grab1 && reader.ready()) {
+                line1 = reader.readLine().trim();
+                if (line1.startsWith("#") || line1.length() == 0) continue;
+                grab1 = false;
+            }
+            while (grab2 && reader.ready()) {
+                line2 = reader.readLine().trim();
+                if (line2.startsWith("#") || line2.length() == 0) continue;
+                grab2 = false;
+            }
+            if ((line1 == null) || (line2 == null)) {
+                break;
+            }
+            //
+            String[] inputs  = line1.split(DEFAULT_VECTORDELIMITER);
+            String[] targets = line2.split(DEFAULT_VECTORDELIMITER);
+            //
+            final int inputlength  = inputs.length;
+            final int targetlength = targets.length;
+            
+            double[][] data1 = transform(inputs);
+            double[][] data2 = transform(targets);
+            
+            final int inputsize  = maxSize(data1);
+            final int targetsize = maxSize(data2);
+            
+            final double[] input  = new double[inputlength * inputsize];
+            final double[] target = new double[targetlength * targetsize];
+            //
+            map(data1, input, inputsize);
+            map(data2, target, targetsize);
+            //
+            Sample sample = new Sample(
+                input, target, 
+                inputsize, inputlength, 
+                targetsize, targetlength
+            );
+            set.add(sample);
+        }
+        reader.close();
+        //
+    }
+
+    private static String asSeqString(
+            final double[] data, 
+            final int size, 
+            final int length
+    ) {
+        StringBuilder out = new StringBuilder();
+        //
+        int off = 0;
+        //
+        for (int i = 0; i < length; i++) {
+            if (i > 0) out.append(DEFAULT_VECTORDELIMITER);
+            //
+            for (int j = 0; j < size; j++) {
+                if (j > 0) out.append(DEFAULT_VALUEDELIMITER);
+                final double value = data[off++];
+                out.append(value);
+            }
+        }
+        //
+        return out.toString();
+    }
+    
+    
+    public static void writeCSV(final SampleSet set, final String filename) throws IOException {
+        //
+        File file          = new File(filename);
+        BufferedWriter out = new BufferedWriter(new FileWriter(file));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *out*. The resource is referenced at lines 323, 327, 331, 341, 342, 343, 345. The resource is closed at line 346. There are other execution paths that do not contain closure statements, for example, when Writer.write throws an exception. Ensure *out* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).